ops/testing.py

```diff
@@ -1075,6 +1090,57 @@ def __init__(self, resource_name):
         self.name = resource_name
 
 
+@_record_calls
+class _TestingConfigData(dict):
+    _supported_types = {
+        'string': str,
+        'boolean': bool,
+        'int': int,
+        'float': float
```

Comment:
It seems strange to subclass dict here, but then to change interactions with class instances to use the added "_config_set" method.  As far as I can tell this is just an internal class used only by the backend - so charmers shouldn't even be able to access the `[key]` indexing operation. I think the below error checks can just go directly in `__setitem__`

Reply:
Charmers will be able to access `__setitem__` if they do `self.config[foo]=bar` from charm code. If they do, now, they will get exactly the same error message they'd get in production. Before: it would just work.